examples/advanced/ocus_explanations.py

```diff
@@ -0,0 +1,302 @@
+
+import numpy as np
+
+def frietkot_explain(verbose=False):
+    '''
+        The 'frietkot' problem; invented by Tias to explain SAT and SAT solving
+        http://homepages.vub.ac.be/~tiasguns/frietkot/
+    '''
+    # Construct the model.
+    mayo = boolvar(name="mayo")
+    ketchup = boolvar(name="ketchup")
+    curry = boolvar(name="curry")
+    andalouse = boolvar(name="andalouse")
+    samurai = boolvar(name="samurai")
+
+    # Pure CNF
+    Nora = mayo | ketchup
+    Leander = ~samurai | mayo
+    Benjamin = ~andalouse | ~curry | ~samurai
+    Behrouz = ketchup | curry | andalouse
+    Guy = ~ketchup | curry | andalouse
+    Daan = ~ketchup | ~curry | andalouse
+    Celine = ~samurai
+    Anton = mayo | ~curry | ~andalouse
+    Danny = ~mayo | ketchup | andalouse | samurai
+    Luc = ~mayo | samurai
+
+    allwishes   = [Nora, Leander, Benjamin, Behrouz, Guy, Daan, Celine, Anton, Danny, Luc]
+
+    wish_weights = [40, 40, 60, 60, 60, 60, 20, 60, 80, 40]
+
+    explanation_sequence = explain_ocus(allwishes, wish_weights, hard=[], verbose=verbose)
+
+    if verbose > 0:
+        print("\nExplanation Sequence:")
+        print("-" * len("Explanation Sequence:"))
+        for id, explanation in enumerate(explanation_sequence):
+            print(f"\n{id}. Using Constraints:\t", explanation["constraints"])
+            print("\t     Derived:\t", explanation["derived"])
+            print("\t        Cost:\t", explanation["cost"])
+
+    return explanation_sequence
+
+def explain_ocus(soft, soft_weights=None,  hard=[], solver="ortools", verbose=False):
+    '''
+        A SAT-based explanation sequence generating algorithm using assumption
+        variables and weighted UNSAT core (Optimal Constrained Unsatisfiable Subset)
+        extraction. [1, 2]
+
+        [1] Bogaerts, B., Gamba, E., Claes, J., & Guns, T. (2020). Step-wise explanations
+        of constraint satisfaction problems. In ECAI 2020-24th European Conference on
+        Artificial Intelligence, 29 August-8 September 2020, Santiago de Compostela,
+        Spain, August 29-September 8, 2020-Including 10th Conference on Prestigious
+        Applications of Artificial Intelligence (PAIS 2020) (Vol. 325, pp. 640-647).
+        IOS Press; https://doi. org/10.3233/FAIA200149.
+
+        [2] Bogaerts, B., Gamba, E., & Guns, T. (2021). A framework for step-wise explaining
+        how to solve constraint satisfaction problems. Artificial Intelligence, 300, 103550.
+    '''
+    reified_soft = []
+    ind = BoolVar(shape=len(soft), name="ind")
+
+    ## adding assumption variables to the soft constraints
+    for i,bv in enumerate(ind):
+        reified_soft += [bv.implies(soft[i])]
+
+    # to map indicator variable back to soft_constraints
+    indmap = dict((v,i) for (i,v) in enumerate(ind))
+
+    # adding the assumption variables
+    I = set(bv for bv in ind)
+
+    Iend = optimal_propagate(hard + reified_soft, I, solver, verbose)
```

Comment:
I think this should just be `solution_itnersection(Model(soft))` or smth similar, it does not need the reified formulation?

Iend is a partial assignment right, not necessarily a list/set of Booleans?

hmm.. maybe that is why the set(v == v.value()) worked below, CPMpy for Booleans converts that to BoolVarImpl or NegBoolView, so single objects, but this is not true for integer variables...